File: Labs/Lab 5/IsValidBST.py
class TreeNode(object):

     def __init__(self, val=0, left=None, right=None):

         self.val = val
         self.left = left
         self.right = right
         
     def isValidBST(self, root):
         def isValid(node, min_val, max_val):
             if not node:
                 return True
             
             if not(min_val < node.val < max_val):
                 return False
             
             return isValid(node.left, min_val, node.val) and \
               isValid(node.right, node.val, max_val)
                    
         return isValid(root, float("-inf"), float("inf"))
     
root1 = TreeNode(4);
root1.left = TreeNode(3);
root1.right = TreeNode(8);
root1.left.left = TreeNode(1);
root1.right.left = TreeNode(5);
root1.right.right = TreeNode(9);

root2 = TreeNode(1);
root2.left = TreeNode(2);
root2.right = TreeNode(8);
root2.left.left = TreeNode(3);
root2.left.right = TreeNode(4);
root2.right.left = TreeNode(5);
root2.right.right = TreeNode(6);
     
tree_node = TreeNode()
print(tree_node.isValidBST(root1))
print(tree_node.isValidBST(root2))

File: Labs/Lab 5/IsValidBST.java
class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) { this.val = val; }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
    public class IsValidBST {
        TreeNode previous; // Declare previous here
    
        public boolean isValidBST(TreeNode root) {
            previous = null; // Initialize previous here
            return isValid(root);
        }
    
public boolean isValid(TreeNode node) {
    if (node == null)
        return true;

    // Check left subtree
    if (!isValid(node.left))
        return false;

    // Check current node
    if (previous != null && node.val <= previous.val)
        return false;
    previous = node;

    // Check right subtree
    return isValid(node.right);
}

public static void main(String[] args){
    TreeNode root1 = new TreeNode(4);
    root1.left = new TreeNode(3);
    root1.right = new TreeNode(8);
    root1.left.left = new TreeNode(1);
    root1.right.left = new TreeNode(5);
    root1.right.right = new TreeNode(9);

    TreeNode root2 = new TreeNode(1);
    root2.left = new TreeNode(2);
    root2.right = new TreeNode(8);
    root2.left.left = new TreeNode(3);
    root2.left.right = new TreeNode(4);
    root2.right.left = new TreeNode(5);
    root2.right.right = new TreeNode(6);

    IsValidBST isValidBST = new IsValidBST();
    System.out.println(isValidBST.isValidBST(root1));
    System.out.println(isValidBST.isValidBST(root2));
    }  
}
